widgets/text/text.go

```diff
@@ -0,0 +1,281 @@
+		opts:       opt,
+	}
+}
+
+// Reset resets the widget back to empty content.
+func (t *Text) Reset() {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	t.buff.Reset()
+	t.givenWOpts = newGivenWOpts()
+	t.scroll = newScrollTracker(t.opts)
+	t.lastWidth = 0
+	t.newText = true
+	t.lines = nil
+}
+
+// Write writes text for the widget to display. Multiple calls append
+// additional text. The text cannot control characters (unicode.IsControl) or
+// space character (unicode.IsSpace) other than:
+//   ' ', '\n'
+// Any newline ('\n') characters are interpreted as newlines when displaying
+// the text.
+func (t *Text) Write(text string, wOpts ...WriteOption) error {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	if err := validText(text); err != nil {
+		return err
+	}
+
+	pos := t.buff.Len()
+	t.givenWOpts[pos] = newOptsRange(pos, pos+len(text), newWriteOptions(wOpts...))
+	if _, err := t.buff.WriteString(text); err != nil {
+		return err
+	}
+	t.newText = true
+	return nil
+}
+
+// lineTrimNeeded returns true if the text on this line needs to be trimmed.
+// I.e. if the Text gadget was configured to trim lines, the current point falls
+// outside of the canvas and the current rune doesn't start a new line.
+func (t *Text) lineTrimNeeded(cur image.Point, cvs *canvas.Canvas, r rune) bool {
+	if cur.X < cvs.Area().Dx() || r == '\n' {
+		return false
+	}
+	return !t.opts.wrapAtRunes
+}
+
+// minLinesForMarkers are the minimum amount of lines required on the canvas in
+// order to draw the scroll markers ('⇧' and '⇩').
+const minLinesForMarkers = 3
+
+// draw draws the text context on the canvas starting at the specified line.
+// Argument starts are the starting positions of all the lines in the text.
+func (t *Text) draw(text string, cvs *canvas.Canvas, starts []int, fromLine int) error {
+	var cur image.Point // Tracks the current drawing position on the canvas.
+	lines := len(starts)
+	height := cvs.Area().Dy()
+	optRange := t.givenWOpts.forPosition(0) // Text options for the current byte.
+	startPos := starts[fromLine]
+	for i, r := range text {
+		if i < startPos {
+			continue
+		}
+
+		// Draw the scroll up marker on the first line if there is more text
+		// above the canvas.
+		if cur.Y == 0 && height >= minLinesForMarkers && fromLine > 0 {
+			if err := cvs.SetCell(cur, '⇧'); err != nil {
+				return err
+			}
+			cur = image.Point{0, cur.Y + 1} // Move to the next line.
+			startPos = starts[fromLine+1]   // Skip one line of text, the marker replaced it.
+			continue
+		}
+
+		if r == '\n' || wrapNeeded(r, cur.X, cvs.Area().Dx(), t.opts) {
+			cur = image.Point{0, cur.Y + 1} // Move to the next line.
+		}
+
+		// Draw the scroll down marker on the last line if there is more text
+		// below the canvas.
+		if cur.Y == height-1 && height >= minLinesForMarkers && height < lines-fromLine {
+			if height >= minLinesForMarkers {
+				if err := cvs.SetCell(cur, '⇩'); err != nil {
+					return err
+				}
+			}
+			break
+		}
+
+		if r == '\n' {
```

Comment:
Do we actually need this statement?

Reply:
Yes, otherwise we print out the newline rune below which adds an empty line of text.

Note - I am not happy with this function as it is in this pull request. I have refactored it into something hopefully easier to understand in the next one.